examples/amm/src/contract.rs

```diff
@@ -141,24 +136,110 @@ impl Amm {
                 }
 
                 let balance0 = self.get_pool_balance(0).await?;
                 let balance1 = self.get_pool_balance(1).await?;
 
-                if balance1 > 0
-                    && abs_diff_ne!(
-                        balance0 as f64 / balance1 as f64,
-                        token0_amount as f64 / token1_amount as f64,
-                        epsilon = 0.01
-                    )
-                {
-                    return Err(AmmError::BalanceRatioAlteredError);
+                let token0_amount;
+                let token1_amount;
+                if balance0 > Amount::ZERO && balance1 > Amount::ZERO {
+                    let balance0_bigint = BigUint::from_u128(u128::from(balance0))
+                        .expect("Couldn't generate balance0 in bigint");
+                    let balance1_bigint = BigUint::from_u128(u128::from(balance1))
+                        .expect("Couldn't generate balance1 in bigint");
+                    let max_token0_amount_bigint =
+                        BigUint::from_u128(u128::from(max_token0_amount))
+                            .expect("Couldn't generate max_token0_amount in bigint");
+                    let max_token1_amount_bigint =
+                        BigUint::from_u128(u128::from(max_token1_amount))
+                            .expect("Couldn't generate max_token1_amount in bigint");
+
+                    // This is the formula to maintain the ratio:
+                    //      balance0 / balance1 = (balance0 + max_token0_amount) / (balance1 + token1_amount)
+                    //      balance0 * (balance1 + token1_amount) = balance1 * (balance0 + max_token0_amount)
+                    //      balance0 * balance1 + balance0 * token1_amount = balance1 * balance0 + balance1 * max_token0_amount
+                    //      balance0 * token1_amount = balance1 * max_token0_amount
+                    //      token1_amount = (balance1 * max_token0_amount) / balance0
+                    //
+                    // For token0_amount, it would be this:
+                    //      token0_amount = (balance0 * max_token1_amount) / balance1
+
+                    let token0_amount_bigint = (balance0_bigint.clone()
+                        * max_token1_amount_bigint.clone())
+                        / balance1_bigint.clone();
+                    let token1_amount_bigint = (balance1_bigint.clone()
+                        * max_token0_amount_bigint.clone())
+                        / balance0_bigint.clone();
+
+                    if token0_amount_bigint > max_token0_amount_bigint
+                        && token1_amount_bigint > max_token1_amount_bigint
+                    {
+                        return Err(AmmError::BalanceRatioAlteredError);
+                    }
+
+                    if token0_amount_bigint <= max_token0_amount_bigint
+                        && token1_amount_bigint <= max_token1_amount_bigint
```

Comment:
And in that case it doesn't really matter, since they should be almost the same. I don't think we need to handle that many cases separately.

Why not just compare `max_token0_amount_bigint * balance1_bigint` and `max_token1_amount_bigint * balance0_bigint` right away? If the former is greater, the user specified too much token 0, so we compute `token0_amount` and use `max_token1_amount`. If the latter is bigger, vice-versa.

Reply:
Sounds good